File: ch2/2-3.c
#include <stdio.h>
#include <math.h>
#include <string.h>
#include <ctype.h>

int get_map(char c);
int htoi(char s[]);

int main(void)
{
    // always assume it's a valid hex value
    char s1[] = "0xFF00FF";
    char s2[] = "cc0033";
    char s3[] = "DEAD";

    printf("%s = %d\n", s1, htoi(s1));
    printf("%s = %d\n", s2, htoi(s2));
    printf("%s = %d\n", s3, htoi(s3));
    return 0;
}

int get_map(char c)
{
    char map[] = "0123456789abcdef";
    int pos = 0;
    for (int i = 0; i < (int)strlen(map); i++) {
        if (tolower(c) == map[i]) {
            pos = i;
            break;
        }
    }
    return pos;
}

int htoi(char s[])
{
    int total = 0;
    int with_prefix = 0;
    int i = 0;
    if (s[0] == '0' && (s[1] == 'x' || s[1] == 'X')) {
        i = 2;
        with_prefix = 1;
    }
    for (int j = strlen(s)-1; s[i] != '\0'; i++,j--) {
        total += get_map(s[i]) * pow(16, j - (with_prefix == 1 ? 2 : 0));
    }
    return total;
}

File: ch2/2-3.go
package main

import (
	"fmt"
	"math"
	"strings"
)

func main() {
	s1 := "0xFF00FF"
	s2 := "cc0033"
	s3 := "DEAD"

	fmt.Printf("%s = %d\n", s1, htoi(s1))
	fmt.Printf("%s = %d\n", s2, htoi(s2))
	fmt.Printf("%s = %d\n", s3, htoi(s3))
}

func getMap(c rune) int {
	m := "0123456789abcdef"
	var pos int
	for i := 0; i < len(m); i++ {
		if strings.ToLower(string(c)) == string(m[i]) {
			pos = i
			break
		}
	}
	return pos
}

func htoi(s string) int {
	var total int
	var prefix bool
	var i int
	if s[0] == '0' && (s[1] == 'x' || s[1] == 'X') {
		i = 2
		prefix = true
	}
	for j := len(s) - 1; i <= len(s)-1; i, j = i+1, j-1 {
		var amt int
		if prefix {
			amt = 2
		}
		total += getMap(rune(s[i])) * int(math.Pow(16.0, float64(j-amt)))
	}
	return total
}
